### PR TITLE
Required recipes to build images and toolchains to build WPE

### PR DESCRIPTION
**Pull request by psaavedra**
* Add packagegroup-wpewebkit-depends. Need to build images and toolchains with all the dependencies to build or run WebKit but without WebKit
* wpewebkit: bumb trunk version (compatible with wpe-1.0)
* Add required recipes to build to run the WebKit Layout tests
  * ~~reverted a patch in Ruby recipe. This is needed to build the Ruby packages from meta-cloud-services/meta-openstack~~
  * added json and highline ruby packages
  * added mrobinson's dicts and fonts resources

### CODE REVIEW

File: .gitlab-ci.yml [hunk 1]
```diff
@@ -9,7 +9,7 @@ thud-wandboard-mesa-weston-wpe-2.20:
     - pushd sources/meta-webkit
     - git checkout $CI_COMMIT_SHA
     - popd
-    - source setup-environment wandboard-mesa-wpe-2.20 --update-config
+    - source setup-environment wandboard-mesa-wpe-2.20 wandboard-mesa browsers wandboard mesa-wpe-220 --update-config
```
Comment (philn) [inline .gitlab-ci.yml]: Version numbers without separator can quickly become very confusing. Can you use a separator? - or _ ?

File: recipes-browser/packagegroups/packagegroup-wpewebkit-depends.bb [hunk 2]
```diff
@@ -0,0 +1,281 @@
...
+    python-psutil \
+    psmisc \
+    pulseaudio \
+    pulseaudio-misc \
+    highline \
+    perl \
+    perl-module-file-spec \
+    perl-module-cgi \
+    json \
+    dicts \
```
Comment (clopez) [inline recipes-browser/packagegroups/packagegroup-wpewebkit-depends.bb]: I think this name "dicts" is too much generic. I would instead use something like "test-dicts"

File: recipes-browser/packagegroups/packagegroup-wpewebkit-depends.bb [hunk 3]
```diff
@@ -0,0 +1,281 @@
...
+    fonts \
```
Comment (clopez) [inline recipes-browser/packagegroups/packagegroup-wpewebkit-depends.bb]: Same for "fonts". I would use "test-fonts" maybe or "webkit-test-fonts"

File: recipes-browser/packagegroups/packagegroup-wpewebkit-depends.bb [hunk 4]
```diff
@@ -0,0 +1,281 @@
...
+    patch \
+    procps \
+    psmisc \
+    sed \
+    shadow \
+    tar \
+    time \
+    util-linux \
+    xdg-utils \
+    \
```
Comment (clopez) [inline recipes-browser/packagegroups/packagegroup-wpewebkit-depends.bb]: This extra empty line is not needed

File: recipes-browser/packagegroups/packagegroup-wpewebkit-depends.bb [hunk 5]
```diff
@@ -0,0 +1,281 @@
...
+    sqlite3 \
+    zlib \
+    libpng \
+    libsoup-2.4 \
+    libwebp \
+    libxml2 \
+    libxslt \
+    libepoxy \
+    libgcrypt \
+    \
```
Comment (clopez) [inline recipes-browser/packagegroups/packagegroup-wpewebkit-depends.bb]: Another extra empty line

File: recipes-devtools/ruby/ruby_2.5.3.bbappend [hunk 6]
```diff
@@ -0,0 +1 @@
+SRC_URI_remove = "file://CVE-2018-1000073.patch"
```
Comment (clopez) [inline recipes-devtools/ruby/ruby_2.5.3.bbappend]: Why is needed to remove this patch from ruby?

Comment (psaavedra) [reply recipes-devtools/ruby/ruby_2.5.3.bbappend]: Some recipes from meta-cloud-services are not pretty up to date and fails due to this fix. In particular packages from meta-cloud-services/meta-openstack.

Of course this is a temporary fix done time ago (note this PR is still WIP). Maybe this is not needed anymore.

Comment (psaavedra) [reply recipes-devtools/ruby/ruby_2.5.3.bbappend]: Not needed anymore. Removed.

File: recipes-browser/packagegroups/packagegroup-wpewebkit-depends.bb (hunk 2 above)
Comment (psaavedra) [reply recipes-browser/packagegroups/packagegroup-wpewebkit-depends.bb]: done

File: recipes-browser/packagegroups/packagegroup-wpewebkit-depends.bb (hunk 3 above)
Comment (psaavedra) [reply recipes-browser/packagegroups/packagegroup-wpewebkit-depends.bb]: done

File: recipes-browser/packagegroups/packagegroup-wpewebkit-depends.bb (hunk 4 above)
Comment (psaavedra) [reply recipes-browser/packagegroups/packagegroup-wpewebkit-depends.bb]: done

File: recipes-browser/packagegroups/packagegroup-wpewebkit-depends.bb (hunk 5 above)
Comment (psaavedra) [reply recipes-browser/packagegroups/packagegroup-wpewebkit-depends.bb]: done

File: recipes-devtools/ruby/ruby-json_1.8.6.bb [hunk 7]
```diff
@@ -0,0 +1,79 @@
...
+  system("perl -p -i -e 's#^topdir.*#topdir = ${sysroot_ruby}#' Makefile 2>&1 > /tmp/jjjjjj")
+  system("perl -p -i -e 's#^topdir.*#topdir = ${sysroot_ruby}#' Makefile")
+  system("perl -p -i -e 's#^hdrdir.*#hdrdir = ${sysroot_ruby}#' Makefile")
+  system("perl -p -i -e 's#^arch_hdrdir.*#arch_hdrdir = ${sysroot_ruby}/\\\\\$(arch)#' Makefile")
+  system("perl -p -i -e 's#^arch =.*#arch = ${ruby_arch}#' Makefile")
+  system("perl -p -i -e 's#^LIBPATH =.*#LIBPATH = -L.#' Makefile")
+  system("perl -p -i -e 's#^dldflags =.*#dldflags = ${LDFLAGS}#' Makefile")
+  system("perl -p -i -e 's#^ldflags  =.*#ldflags = -L${STAGING_LIBDIR}#' Makefile")
+EOF
+}
```
Comment (clopez) [inline recipes-devtools/ruby/ruby-json_1.8.6.bb:76]: This looks quite hackish.
Wouldn't it be better to do this as a patch to the recipe?

Comment (psaavedra) [reply recipes-devtools/ruby/ruby-json_1.8.6.bb]: A part of some nits that I forgot to remove, those functions are strongly based on the `ruby.class`:  [meta-cloud-services - L110](https://git.yoctoproject.org/cgit/cgit.cgi/meta-cloud-services/tree/classes/ruby.bbclass#n110) and  [meta-cloud-services - L88](https://git.yoctoproject.org/cgit/cgit.cgi/meta-cloud-services/tree/classes/ruby.bbclass#n88)

What can I do is propose those changes to `meta-cloud-services`: https://lists.yoctoproject.org/g/meta-virtualization/message/5294

Comment (clopez) [reply recipes-devtools/ruby/ruby-json_1.8.6.bb]: I see. Interesting hacks coming from meta-cloud-services.
I wonder if a new dependency on any external layer (like meta-cloud-services) its required in order to build this?
I also wonder how the `inherit ruby` statement in the recipes gets resolved? I don't see where it is the `ruby.bbclass` file (its not on poky/oe-core or meta-openembedded as far as I can see)

**Comment by psaavedra** [comment]
On Thu, Apr 30, 2020, 03:48 Carlos Alberto Lopez Perez <
notifications@github.com> wrote:

> *@clopez* commented on this pull request.
> ------------------------------
>
> In recipes-devtools/ruby/ruby-json_1.8.6.bb
> <https://github.com/Igalia/meta-webkit/pull/83#discussion_r417710207>:
>
> > +ruby_do_compile() {
> +        EXTCONF_FILES=$(find . -name extconf.rb -exec ls {} \;)
> +        for e in $EXTCONF_FILES
> +        do
> +            if [ -f $e -a ! -f $e.orig ] ; then
> +                grep create_makefile $e > append2 || (exit 0)
> +                ruby_gen_extconf_fix
> +                cp $e $e.orig
> +                # Patch extconf.rb for cross compile
> +                cat append >> $e
> +            fi
> +        done
> +        for gem in ${RUBY_BUILD_GEMS}; do
> +                ${RUBY_COMPILE_FLAGS} gem build $gem
> +        done
> +        for e in $EXTCONF_FILES
> +        do
> +            if [ -f $e.orig ] ; then
> +                cat $e
> +                mv $e.orig $e
> +            fi
> +        done
> +}
> +
> +
> +ruby_gen_extconf_fix() {
> +        cat<<EOF>append
> +  RbConfig::MAKEFILE_CONFIG['CPPFLAGS'] = ENV['CPPFLAGS'] if ENV['CPPFLAGS']
> +  \$CPPFLAGS = ENV['CPPFLAGS'] if ENV['CPPFLAGS']
> +  RbConfig::MAKEFILE_CONFIG['CC'] = ENV['CC'] if ENV['CC']
> +  RbConfig::MAKEFILE_CONFIG['LD'] = ENV['LD'] if ENV['LD']
> +  RbConfig::MAKEFILE_CONFIG['CFLAGS'] = ENV['CFLAGS'] if ENV['CFLAGS']
> +  RbConfig::MAKEFILE_CONFIG['CXXFLAGS'] = ENV['CXXFLAGS'] if ENV['CXXFLAGS']
> +EOF
> +        cat append2>>append
> +        sysroot_ruby=${STAGING_INCDIR}/ruby-${RUBY_GEM_VERSION}
> +        ruby_arch=`ls -1 ${sysroot_ruby} |grep -v ruby |tail -1 2> /dev/null`
> +        cat<<EOF>>append
> +  system("perl -p -i -e 's#^topdir.*#topdir = ${sysroot_ruby}#' Makefile 2>&1 > /tmp/jjjjjj")
> +  system("perl -p -i -e 's#^topdir.*#topdir = ${sysroot_ruby}#' Makefile")
> +  system("perl -p -i -e 's#^hdrdir.*#hdrdir = ${sysroot_ruby}#' Makefile")
> +  system("perl -p -i -e 's#^arch_hdrdir.*#arch_hdrdir = ${sysroot_ruby}/\\\\\$(arch)#' Makefile")
> +  system("perl -p -i -e 's#^arch =.*#arch = ${ruby_arch}#' Makefile")
> +  system("perl -p -i -e 's#^LIBPATH =.*#LIBPATH = -L.#' Makefile")
> +  system("perl -p -i -e 's#^dldflags =.*#dldflags = ${LDFLAGS}#' Makefile")
> +  system("perl -p -i -e 's#^ldflags  =.*#ldflags = -L${STAGING_LIBDIR}#' Makefile")
> +EOF
> +}
>
> I see. Interesting hacks coming from meta-cloud-services.
> I wonder if a new dependency on any external layer (like
> meta-cloud-services) its required in order to build this?
> I also wonder how the inherit ruby statement in the recipes gets
> resolved? I don't see where it is the ruby.bbclass file (its not on
> poky/oe-core or meta-openembedded as far as I can see)
>

No is not there. Ruby.class is part if the meta-cloud-services [1]. I
checked this and yes this is the only reason why I'm using
meta-cloud-services.



https://git.yoctoproject.org/cgit/cgit.cgi/meta-cloud-services/tree/classes/ruby.bbclass

**Comment by clopez** [comment]
> No is not there. Ruby.class is part if the meta-cloud-services [1]. I checked this and yes this is the only reason why I'm using meta-cloud-services. https://git.yoctoproject.org/cgit/cgit.cgi/meta-cloud-services/tree/classes/ruby.bbclass

Two questions:
* I don't see a new dependecy for meta-cloud-services added on this PR. If this new recipes depend on meta-cloud-services then it should be stated somewhere (in the README or in the recipe itself) Something similar to what we did with [meta-python2 and browserfrunner](https://github.com/Igalia/meta-webkit/commit/4bbb1de9d165c7bdffcd966222442724104dffb8) may work.
 * Assuming the new dependency its needed, then it's not clear to me why we are adding recipes to meta-webkit already available on meta-cloud-services like this ruby-json one

**Comment by psaavedra** [comment]
The cross-compiling for rubygems is quite broken nowadays (https://github.com/rubygems/rubygems/issues/1840) this why I suppose the ruby.class does those tricky things.

**Comment by psaavedra** [comment]
> > No is not there. Ruby.class is part if the meta-cloud-services [1]. I checked this and yes this is the only reason why I'm using meta-cloud-services. https://git.yoctoproject.org/cgit/cgit.cgi/meta-cloud-services/tree/classes/ruby.bbclass
> 
> Two questions:
> 
> * I don't see a new dependecy for meta-cloud-services added on this PR. If this new recipes depend on meta-cloud-services then it should be stated somewhere (in the README or in the recipe itself) Something similar to what we did with [meta-python2 and browserfrunner](https://github.com/Igalia/meta-webkit/commit/4bbb1de9d165c7bdffcd966222442724104dffb8) may work.
> * Assuming the new dependency its needed, then it's not clear to me why we are adding recipes to meta-webkit already available on meta-cloud-services like this ruby-json one. (I see it on meta-cloud-services/meta-openstack/recipes-devtools/ruby/json_git.bb)


Yeah, my bad to don't check this part of the MR before. It was created time ago and in use here in [meta-perf-browser](https://gitlab.com/browsers/meta-perf-browser/pipelines) where the meta-cloud-services was added in the bblayers. I totally forget about that until now.

**Comment by psaavedra** [comment]
One idea is just put the ruby.class here in `meta-webkit` and avoid the dependency with another layout. Actually , the target of the meta-cloud-services layers doesn't match very well with the purpose of meta-webkit. WDTY?

**Comment by clopez** [comment]
> One idea could be just to put the ruby.class here in `meta-webkit` and avoid the dependency with another layer. Actually, the target of the `meta-cloud-services` layer doesn't match very well with the purpose of `meta-webkit`. WDTY? (I pushed those changes)

That sounds fine to me

**Comment by psaavedra** [comment]
> > One idea could be just to put the ruby.class here in `meta-webkit` and avoid the dependency with another layer. Actually, the target of the `meta-cloud-services` layer doesn't match very well with the purpose of `meta-webkit`. WDTY? (I pushed those changes)
> 
> That sounds fine to me

So, WDYT, ready for merge?

**Pull request merged**